vyper/semantics/types/function.py

```diff
@@ -314,13 +335,19 @@ def from_FunctionDef(cls, funcdef: vy_ast.FunctionDef) -> "ContractFunctionT":
                     "Default function may not receive any arguments", funcdef.args.args[0]
                 )
 
+        if function_visibility == FunctionVisibility.DEPLOY and funcdef.name != "__init__":
+            raise FunctionDeclarationException(
+                "Only constructors can be marked as `@deploy`!", funcdef
+            )
         if funcdef.name == "__init__":
-            if (
-                state_mutability in (StateMutability.PURE, StateMutability.VIEW)
-                or function_visibility == FunctionVisibility.INTERNAL
-            ):
+            if state_mutability in (StateMutability.PURE, StateMutability.VIEW):
                 raise FunctionDeclarationException(
-                    "Constructor cannot be marked as `@pure`, `@view` or `@internal`", funcdef
+                    "Constructor cannot be marked as `@pure` or `@view`", funcdef
+                )
+            if function_visibility != FunctionVisibility.DEPLOY:
```

Comment:
Currently, the sequence of visibility modifier matter, e.g.:

```vy
@internal
@deploy
def __init__():
    pass
```

Will raise with:

```vy
vyper.exceptions.FunctionDeclarationException: Visibility is already set to: internal
```

I think a better error would be in that particular case:

```vy
vyper.exceptions.FunctionDeclarationException: Constructor must be only marked as `@deploy`, and cannot be `@internal`
```

Reply:
https://github.com/vyperlang/vyper/pull/3729/commits/524a7e86f144f339f8e2b0e78fa8c8ece6b62786

now the error message looks like
```vyper
vyper.exceptions.FunctionDeclarationException: Visibility is already set to: internal

  (hint: only one visibility decorator is allowed per function)

  contract "tmp/repro.vy:2", function "__init__", line 2:1 
       1 @internal
  ---> 2 @deploy
  --------^
       3 def __init__():

```